lib/game.rb

```diff
@@ -54,8 +55,59 @@ def select_difficulty_level
     ###
   end
 
-  def select_ship_attributes
-    default_attributes = [['Cruiser',3],['Submarine',2]]
+  def select_ship_attributes(board_area, rows, columns)
+
+    @message.would_you_like_to_customize_ships?
+    valid_selections = [1, 2, 3, 4]
+    ship_selection = gets.chomp.to_i
+
+    until valid_selections.include?(ship_selection)
+      @message.invalid_customize_ship_selection
+      ship_selection = gets.chomp.to_i
+    end
+
+    # Ship lengths must be < 1/3 of the board area to play
+    case ship_selection
+    when 1
+      if (board_area / 3) < 5
+        @message.board_is_too_small_for_ship_selection
+        return select_ship_attributes(board_area)
+      end
+      return [['Cruiser', 3],['Submarine', 2]]
+    when 2
+      if (board_area / 3) < 9
+        @message.board_is_too_small_for_ship_selection
+        return select_ship_attributes(board_area)
+      end
+      return [['Battleship', 4],['Cruiser', 3],['Submarine', 2]]
+    when 3
+      if (board_area / 3) < 14
+        @message.board_is_too_small_for_ship_selection
+        return select_ship_attributes(board_area)
+      end
+      return [['Carrier', 5],['Battleship', 4],['Cruiser', 3],['Submarine', 2]]
+    when 4
+      return user_custom_ships(board_area, rows, columns)
+    end
+
+  end
+
+  def user_custom_ships(board_area, rows, columns)
+    user_input = "Y"
+    user_ships = []
+
+    until user_input == "N"
+      user_ships_total_length = user_ships.map { |ship| ship[1] }.inject(0){ |sum, x| sum + x }
+      ship_name  = @message.prompt_user_for_custom_ship_name(user_ships, user_ships_total_length, board_area)
+      ship_length = @message.prompt_user_for_custom_ship_length(ship_name, user_ships_total_length, board_area, rows, columns)
+
+      user_ships << [ship_name, ship_length]
+
+      user_input = @message.succusfully_created_a_ship(ship_name, ship_length)
```

Comment:
Should there be a check in this loop (Maybe in the until?) to ensure that it would be possible for the user to add another ship (of minimum length two or maybe one)

Reply:
Yes, I did not think of this. Especially for smaller boards, the user could create a ship that pushes the "user_ships_total_length" above "board_area / 3".  And now they are stuck in an infinite loop.  I'll work on a fix

Reply:
Added to the until loop: || (user_ships_total_length + 2) > board_area / 3
This will check that a 2 length ship does not push the total lengths > board_area / 3

I still see an issue that we are not checking if there exists 2 consecutive cells.  But I think this would have to be a really small board for that to ever happen if two thirds of the board cells are still available.